flow/Trace.cpp

```diff
@@ -220,6 +221,46 @@ struct TraceLog {
 		}
 	};
 
+	struct IssuesList : ITraceLogIssuesReporter, ThreadSafeReferenceCounted<IssuesList> {
+		IssuesList(){};
+		void addAndExpire(std::string issue, double expirationInterval) override {
+			MutexHolder h(mutex);
+			auto now = ::now();
+			if (issues.find(issue) != issues.end()) {
+				issues[issue]++;
+			} else {
+				issues[issue] = 1;
+			}
+			queue.emplace_back(now + expirationInterval, issue);
+		}
+
+		void retrieveIssues(std::vector<std::string>& out) override {
+			MutexHolder h(mutex);
+			// clean up any expired events first
+			auto now = ::now();
+			while (queue.size() > 0 && queue.front().first <= now) {
+				ASSERT(issues.find(queue.front().second) != issues.end());
+				if (--issues[queue.front().second] == 0) {
+					issues.erase(queue.front().second);
```

Comment:
This is doing 3 map lookups into issues, but I think it could be 1 if written like:

```
auto itr = issues.find(queue.front().second);
ASSERT(itr != issues.find.end);
if(--itr->second == 0) {
	issues.erase(itr);
}
```

Reply:
Done